src/PasswordHashingProvider.cpp

```diff
@@ -0,0 +1,63 @@
+#include "./PasswordHashingProvider.hpp"
+#include "../external/argon2/include/argon2.h"
+#include <pthread.h>
+#include <iomanip>
+#include <random>
+#include <memory>
+#include <algorithm>
+#include <iostream>
+#include <sstream>
+
+PasswordHashingProvider::PasswordHashingProvider() : 
+    PasswordHashingProvider(16, 16)
+{}
+
+PasswordHashingProvider::PasswordHashingProvider(unsigned pwdSaltLength, unsigned pwdHashLength) :
+    passwordSaltLength(pwdSaltLength), passwordHashLength(pwdHashLength) {
+    checkAndCorrectInitialValues(pwdSaltLength, pwdHashLength);
+}
+
+PasswordHashingProvider::~PasswordHashingProvider() {
+}
+
+void PasswordHashingProvider::printSaltAndHashLength() const {
+    std::cout << "PasswordHashingProvider instance:\n";
+    std::cout << "  salt length:" << passwordSaltLength << '\n';
+    std::cout << "  hash length:" << passwordHashLength << '\n';
+    std::cout << "  iterations:" << iterations << '\n';
+    std::cout << "  memory cost:" << memoryCost << '\n';
+    std::cout << "  threads:" << threads << '\n';
+}
+
+std::string PasswordHashingProvider::generateSalt() const {
+    std::random_device randDev;
+    std::mt19937 randEng(randDev());
+    std::uniform_int_distribution<unsigned> uIDistr(0,9);  
+    std::stringstream stringStream{};
+    for (unsigned it = 0; it < passwordSaltLength; ++it) 
+        stringStream << uIDistr(randEng);
+    return stringStream.str();
+}
+
+std::string PasswordHashingProvider::generateHash(const std::string& passwordSalt, const std::string& passwordPhrase) const {
+    std::unique_ptr<uint8_t[]> temporaryHash = std::make_unique<uint8_t[]>(passwordHashLength);
+    argon2d_hash_raw(
+        iterations, memoryCost, threads, passwordPhrase.c_str(), passwordPhrase.length(), 
+        passwordSalt.c_str(), passwordSalt.length(), temporaryHash.get(), passwordHashLength);
```

Comment:
hmm... including whole library only for this one quite complicated function may be a little bit overkill. 
What do you think about https://en.cppreference.com/w/cpp/string/basic_string/hash?

Reply:
I am aware the fact that using external library for hashing passwords is an overkill for our project. Nonetheless I have choosed to use it for a set of reasons:
- to gain skills with linking static libraries in CMake project and writing own code in general;
- according to [cppReference](https://en.cppreference.com/w/cpp/utility/hash) std::hash function is using hashing algorithm (murmurhash family) that is not intended for password hashing due to security reasons;
- libargon2 library was made strictly for password hashing, has reasonable size and as far as I know, it does not implement any other functionalities.

Reply:
Very good explanation. Thanks :)
You convinced me that it can stay here. But please try to remove vs2015 directory from libargon, which seems to be not needed.